gcsfs/core.py

```diff
@@ -574,28 +581,149 @@ async def _list_objects(self, path, prefix="", versions=False):
             else:
                 return []
         out = pseudodirs + items
-        # Don't cache prefixed/partial listings
-        if not prefix:
+        
+        use_snapshot_listing = False if not inventory_report_info \
+            else inventory_report_info.get("use_snapshot_listing")
+
+        # Don't cache prefixed/partial listings, in addition to 
+        # not using the inventory report service to do listing directly.
+        if not prefix and use_snapshot_listing is False:
             self.dircache[path] = out
         return out
 
     async def _do_list_objects(
-        self, path, max_results=None, delimiter="/", prefix="", versions=False
-    ):
+        self, path, max_results=None, delimiter="/", 
+                    prefix="", versions=False, **kwargs):
+        
         """Object listing for the given {bucket}/{prefix}/ path."""
         bucket, _path, generation = self.split_path(path)
         _path = "" if not _path else _path.rstrip("/") + "/"
         prefix = f"{_path}{prefix}" or None
 
+        # Page size of 5000 is officially supported across GCS.
+        default_page_size = 5000
+
+        inventory_report_info = kwargs.get("inventory_report_info", None)      
+
+        # Check if the user has configured inventory report option.
+        if inventory_report_info is not None:
+            items, prefixes = await \
+                InventoryReport.fetch_snapshot(
+                    gcs_file_system=self,
+                    inventory_report_info=inventory_report_info,
+                    prefix=prefix)
+            
+            use_snapshot_listing = inventory_report_info.get("use_snapshot_listing")
+
+            # If the user wants to rely on the snapshot from the inventory report
+            # for listing, directly return the results.
+            if use_snapshot_listing:
+                return items, prefixes
+            
+            # Otherwise, use the snapshot to initate concurrent listing.
+            return await self._concurrent_list_objects_helper(
+                items=items,
+                bucket=bucket,
+                delimiter=delimiter,
+                prefix=prefix,
+                versions=versions,
+                generation=generation,
+                page_size=default_page_size
+            )
+        
+        # If the user has not configured inventory report, proceed to use
+        # sequential listing.
+        else:
+            return await self._sequential_list_objects_helper(
+                bucket=bucket,
+                delimiter=delimiter,
+                start_offset=None,
+                end_offset=None,
+                prefix=prefix,
+                versions=versions,
+                generation=generation,
+                page_size=default_page_size)
+        
+    async def _concurrent_list_objects_helper(self, items, bucket, delimiter,
```

Comment:
Sorry if I've missed it, could you point me to where the tests for `_concurrent_list_objects_helper` and `_sequential_list_objects_helper` are added?

Reply:
since `_concurrent_list_objects_helper` and `_sequential_list_objects_helper` are both helper methods in `core.py`, there is currently no explicit testing for the two methods. I was looking at `core.py`, and I see that the other methods are not being unit tested, but rather integration tested. 

There is the `test_inventory_report_listing` that does a simple integration test with the inventory report passed in. As for  `_sequential_list_objects_helper`, the function is basically the original `do_list_objects`, so I personally felt there is no need to test it again.

Let me know your thoughts @Magichan33.

Reply:
I see. Sounds good to me! I'll let Martin have the final say on this though.

Reply:
Where the higher function simply dispatches to two helper functions, it is fine not to write separate tests of the helper functions. The exception would be to make sure that invalid sets of arguments raise correctly, where the higher function should never send such arguments. I don't think this applies here.